src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -293,115 +376,306 @@ public bool TryGetValue(string key, out object value)
-        {
-            private readonly RouteValueDictionary _dictionary;
+            public override int Count => _inner.Count;
 
-            private int _index;
-            private Dictionary<string, object>.Enumerator _enumerator;
+            public override KeyValuePair<string, object> this[int index] => _inner[index];
 
-            public Enumerator(RouteValueDictionary dictionary)
+            public void Clear()
             {
-                if (dictionary == null)
+                _inner.Clear();
+            }
+
+            public override bool ContainsKey(string key)
+            {
+                for (var i = 0; i < _inner.Count; i++)
                 {
-                    throw new ArgumentNullException();
+                    var kvp = _inner[i];
+                    if (string.Equals(key, kvp.Key, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return true;
+                    }
                 }
 
-                _dictionary = dictionary;
+                return false;
+            }
 
-                Current = default(KeyValuePair<string, object>);
-                _index = -1;
-                _enumerator = _dictionary.InnerDictionary == null ?
-                    default(Dictionary<string, object>.Enumerator) :
-                    _dictionary.InnerDictionary.GetEnumerator();
+            public override bool TrySetValue(string key, object value)
+            {
+                for (var i = 0; i < _inner.Count; i++)
+                {
+                    var kvp = _inner[i];
+                    if (string.Equals(key, kvp.Key, StringComparison.OrdinalIgnoreCase))
+                    {
+                        _inner[i] = new KeyValuePair<string, object>(key, value);
+                        return true;
+                    }
+                }
+
+                _inner.Add(new KeyValuePair<string, object>(key, value));
+                return true;
             }
 
-            public KeyValuePair<string, object> Current { get; private set; }
+            public override bool TryGetValue(string key, out object value)
+            {
+                for (var i = 0; i < _inner.Count; i++)
+                {
+                    var kvp = _inner[i];
+                    if (string.Equals(key, kvp.Key, StringComparison.OrdinalIgnoreCase))
+                    {
+                        value = kvp.Value;
+                        return true;
+                    }
+                }
 
-            object IEnumerator.Current => Current;
+                value = null;
+                return false;
+            }
 
-            public void Dispose()
+            public override void Upgrade(ref Storage storage)
             {
+                // Do nothing.
             }
+        }
 
-            public bool MoveNext()
+        internal class PropertyStorage : Storage
+        {
+            private static readonly PropertyCache _propertyCache = new PropertyCache();
```

Comment:
Ick, another `static` cache on top of those in `PropertyHelper`. Any way we can avoid that? E.g. could expose a `virtual ValidateHelpers(PropertyHelper[] helpers)` method that `PropertyHelper.GetProperties(...)` calls before updating its cache.